cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -427,8 +427,38 @@ void ShiftTime(FunctionContext* ctx, const CastOptions& options, const bool is_m
       out_data[i] = static_cast<out_type>(in_data[i]);
     }
   } else if (is_multiply) {
-    for (int64_t i = 0; i < input.length; i++) {
-      out_data[i] = static_cast<out_type>(in_data[i] * factor);
+    if (options.allow_time_truncate) {
+      for (int64_t i = 0; i < input.length; i++) {
+        out_data[i] = static_cast<out_type>(in_data[i] * factor);
+      }
+    } else {
+#define RAISE_OVERFLOW_CAST(VAL)                                                  \
+  ctx->SetStatus(Status::Invalid("Casting from ", input.type->ToString(), " to ", \
+                                 output->type->ToString(), " would result in ",   \
+                                 "out of bounds timestamp: ", VAL));
+
+      if (input.null_count != 0) {
+        internal::BitmapReader bit_reader(input.buffers[0]->data(), input.offset,
```

Comment:
Using `ArrayDataVisitor` would allow abstracting this away.

Reply:
@pitrou I tries doing this, for now without taking into the options or templating, with:

```cpp
// <TimestampType, TimestampType> and <DurationType, DurationType>
template <typename O, typename I>
struct CastFunctor<O, I, ..enable..>
  void operator()(FunctionContext* ctx, const CastOptions& options,
                  const ArrayData& input, ArrayData* output) {
    // If units are the same, zero copy, otherwise convert
    const auto& in_type = checked_cast<const I&>(*input.type);
    const auto& out_type = checked_cast<const O&>(*output->type);

    if (in_type.unit() == out_type.unit()) {
      ZeroCopyData(input, output);
      return;
    }

    // std::pair<bool, int64_t> conversion =
    //     kTimeConversionTable[static_cast<int>(in_type.unit())]
    //                         [static_cast<int>(out_type.unit())];

    in_data_ = input.GetValues<int64_t>(1);
    out_data_ = output->GetMutableValues<int64_t>(1);

    ArrayDataVisitor<I> visitor;
    Status st = visitor.Visit(input, this);
    if (!st.ok()) {
      ctx->SetStatus(st);
      return;
    }
    // ShiftTime<int64_t, int64_t>(ctx, options, conversion.first, conversion.second, input,
    //                             output);
  }

  Status VisitNull() {
    idx_ += 1;
    return Status::OK();
  }

  Status VisitValue(int64_t value) {
    out_data_[idx_] = static_cast<int64_t>(value * 1000);
    idx_ += 1;
    return Status::OK();
  }

  int64_t idx_ = 0;
  const int64_t* in_data_;
  int64_t* out_data_;
};
```

Is keeping track of an index like that a good way to write to `out_data_` ? (the `VisitValue` is not aware of an index I think) Another option could be to memcpy to it and shift it each time?

But from a quick timing, this seems to be quite a bit slower as the specialized loops in `ShiftTime` (for an array of 100k elements it went from 240 us to 1.6 ms) That's not expected? Or is there something I am doing wrong in the above implementation?

Reply:
Writing `out_data_` looks fine to me.

As for `ShiftTime`, it's optimal as it is now: it can operate without checking for null bits, which is faster. The case where you have to check for null bits (because you shouldn't try to detect an overflow on a null value) is where `ArrayDataVisitor` is recommended.

Reply:
> The case where you have to check for null bits (because you shouldn't try to detect an overflow on a null value) is where ArrayDataVisitor is recommended.

Also for that case, the current code is quite a bit more efficient (when a null is introduced, casting goes from ca 300us to 640us using the existing `ShiftTime` implementation, while the ArrayDataVisitor (as implemented above) gives ca 1.6ms)

Reply:
Interesting. Are you sure you measured in release mode?

Reply:
Ah, that's a good question. I did not, sorry.

And with release mode, everything is much faster, and using the ArrayDataVisitor does not give any slowdown (also not for the non-null case it seems). 
OK, will look a bit further into integrating all the options with the code snippet above.